### PR TITLE
feat: show running builds in the Debug page

### LINKED ISSUE #25759: Debug 1.1 - Show Running Build details
### Current behavior
On the Debug page, if the only runs found that match any of the commit shas passed to the Cloud are RUNNING, then the empty state that shows the count of specs that have completed is shown  and updated every 15 seconds.  When the build completes, the empty state spec count is replaced with either a success message if no failed tests are reported, or else the failed tests are displayed.

image

On the Debug page, if a completed build is found and shown on the page, and a newer run is found with the RUNNING status, a banner is shown at the top of the page just below the header that shows the count of completed and total specs for the RUNNING build and is updated every 15 seconds.  When that build completes, the banner updates to show a "View run" link.  Clicking that link will update the Debug page to show the details for that completed run.

image

image

### Desired behavior

---
#### Testing in progress banner
On the Debug page, if the current run is in a RUNNING status, replace the empty state with just the count of specs completed vs total with the following:
* Show a banner at the top of the page that contains the run progress with the following states:
  * While tests are being recorded
    * Show copy  "Testing in progress..." with the spec counts "X of X spec completed"
    * Show progress bar with percent complete calculated by the ratio of specs completed vs total
    * Right most part of the progress bar will have a different color background to indicate the "delayed completion" stage of a run.  Hovering over this will show a tooltip (see copy below)

image

  * When run enters the "waiting for completion" phase
    * Add additional message that says "Scheduled to complete in X seconds".  The seconds should be calculated from the new XXXX timestamp field.  The seconds should dynamically count down to 0.
    * If the time to completion passes (i.e. the countdown reaches 0), change the copy to say "Schedule to complete soon".  This should be rare or only while waiting on a Cloud query to return, etc.

**NOTE** Removing from the scope for this issue the UI element in the progress bar shown below and the tooltip
image

_Delayed completion tooltip copy_
* Title
  * Run completion delay
* Body
  * Run will wait for new groups to join before transitioning to 'completed'. [Edit delay in project settings](http://url_to_project_settings)

---
#### New empty state copy while running
* Show empty state copy when no failed tests have been returned
  * Update `DebugPendingRunSpash` to have a title of "Testing in progress..." and a sub title of "Failures will be displayed here"

image

* If the run enters the "scheduled to complete" phase and there are no failures, the empty state copy should update to say "Scheduled to complete..."

image

---
#### Showing failed tests while RUNNING
* Show failed tests while the run is RUNNING
  * Allow the debug container to show the failed tests, even when the run is in the RUNNING status
  * Update the poller in `RelevantRunSpecsDataSource` to track count of failed tests and emit on the `relevantRunChange` subscription if the test counts change.  This will cause the GQL query in the Debug page to refetch.
  * When new data is returned from the Cloud that should update what failed tests are shown, these should animate in so that the change in the page is not janky.  Updates may be needed in `DebugContainer` in the way the data is mapped in the computed value `debugSpecsArray` to support this.
    * Animations 
      * Make use of Vue's [Transition](https://vuejs.org/guide/built-ins/transition.html) and [TransitionGroup](https://vuejs.org/guide/built-ins/transition-group.html#transitiongroup) if possible.  The desire is to do the animations with CSS and not rely on javascript. You can make use of the "move" part of the TransitionGroup to allow them to change position if new ones sort earlier then previously shown ones.
      * New spec
      * New test(s) within a spec
      * Spec changing from 1 group to multiple groups
      * Adding an additional group to a spec that already has multiple groups

---
#### Add a way to switch to a running run

* Add a "Switch to run" link to the `DebugNewRelevantRunBar` when status is RUNNNIG
  * Allow a user to switch seeing a RUNNING run from a completed run.  Clicking the view run button should use the same mutation to switch the current run to the newest run in the RUNNING state

### References
Figma - https://www.figma.com/file/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?node-id=23626%3A225092&t=gSNW0CyMgqjwNqp0-1
Figma animation prototype - https://www.figma.com/proto/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?node-id=24179%3A256286&scaling=min-zoom&page-id=21010%3A176355&starting-point-node-id=24179%3A256286&show-proto-sidebar=1

### PR DESCRIPTION
**Pull request by warrensplayer**
- Closes https://github.com/cypress-io/cypress/issues/25759

### Additional details

Before this update, a developer would have to wait for a run to complete before being able to see the failed tests that may be a part of the run.  This change will allow the developer to switch to a running run and see any failed tests as they get recorded to the Cloud run.

#### Updates

* Updated `DebugNewRelevantRunBar`
  * Was moved above the Debug header when it is shown
  * Link for "View run" was changed to "Switch to run"
  * Link for "Switch to run" was added for a RUNNING run

_Before_
image
image

_After_
Screenshot 2023-03-01 at 11 42 05 AM
Screenshot 2023-03-01 at 11 41 52 AM

* Added `DebugTestingProgress` component
  * Shows when viewing a RUNNING run just below the header
  * Indicates how many specs have completed with text as well as a progress bar
  * Has a countdown when the run has a `completionScheduledAt` timestamp
  * Switches to say completion is soon if countdown expires

Screenshot 2023-03-01 at 11 45 36 AM
Screenshot 2023-03-01 at 11 45 46 AM
Screenshot 2023-03-01 at 11 46 07 AM

_NOTE_ I have not implemented the "Run Completion Delay" tooltip and indication at the end of the progress bar.  I need to review with Product to see if we want to do this in a fast-follow

* Updated `DebugPendingRunSplash`
  * Removes the spec count because they are included in the `DebugTestingProgress` component
  * Update copy to support run is currently processing specs and has a state when the run is scheduled to complete

Screenshot 2023-03-01 at 11 48 12 AM
Screenshot 2023-03-01 at 11 48 27 AM

* Update `DebugPageHeader`
  * Allow for the duration field to count up for a running run

* Updated Sidebar badge to show when status of run is RUNNING

* Updates to support animating the changing parts of the UI
  * Updated the logic `Debug.vue` to cache and replace the data being passed to the `DebugContainer` to make sure the page does not flash when the query is rerunning to update the page
  * Added various `Transition` and `TransitionGroup` components to fade components in and out
  * Modified CSS and the structure of some components
    * Pulled internal padding from some components up to a parent container
    * Updated `StatsMetadata` to support a "prefix" slot so that the render in `DebugSpec` did not render a list (`ul`) within another list using incorrect DOM

* Backend updates
  * Updated `moveToNext` mutation to move to a specific run
  * Updates to `RelevantRunSpecsDataSource`
    * Track overall test counts to know when to trigger a running run to refresh
    * Track new cloud field called `scheduledToCompleteAt` to know when to switch the page to the "scheduled to complete" state
 
#### Extra credit
* Discovered a missing browser icon in `LayeredBrowserIcons` and added a placeholder.  Filed a ticket to get the real icon added.
* Update `useDurationFormat` to show milliseconds if the duration is less than 1 second

### Steps to test

* Component tests to review
  * `packages/app/src/debug/DebugContainer.cy.tsx`
    * Added "renders running run with failed tests" test to some what mimic the phases of a running run
      * NOTE: Does not mock out DebugTestingProgress component due to issues with different mocked Urql queries interfering with each other
  * `packages/app/src/debug/DebugNewRelevantRunBar.cy.tsx`
  * `packages/app/src/debug/DebugPageHeader.cy.tsx`
  * `packages/app/src/debug/DebugPendingRunSplash.cy.tsx`
  * `packages/app/src/debug/DebugTestingProgress.cy.tsx`
  * `packages/app/src/debug/StatsMetadata.cy.tsx`
  * `packages/app/src/navigation/SidebarNavigation.cy.tsx`

* E2E tests
  * The `debug.cy.ts` tests started to fail after including the `scheduledToCompleteAt` field in the query from `DebugContainer`.  When running locally, the page would thrash between states.  I believe the cause was the mocked up responses in `packages/app/cypress/fixtures/debug-Passing/gql-Debug.json` and `packages/app/cypress/fixtures/debug-Failing/gql-Debug.json` not containing this field.  Thus, Urql which see that it was not present and requery causing other queries to also rerun.  After adding the `scheduledToCompleteAt` field, the tests started passing again.
  * For this ticket, I relied on the CT tests for testing out the various states of the components and then live testing against a Cloud instance (staging and production).  I worked on adding a new E2E tests for the running state, but could not get it working in short order so left it out.  

* Options for real world testing
  * Setup a project on a local Cloud instance, on staging, or in production and manually record a run to it
  * Check out the branch for this PR and trigger a Circle CI run against it to watch it.  

### How has the user experience changed?

**See screenshots above**

### PR Tasks

- [x] Have tests been added/updated?
- [-] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [-] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by astone123** [comment]
Wow, this is working really well for me while testing against staging. One thing I'm noticing in initial testing is that there's some flickering back and forth with the navigation badge (seemingly when new data comes in from the cloud?) - I'll see the number of failures, then the badge will flash to "New" for a second, and then back to the number of failures (this time updated since more tests are failing).

We are [removing the New badge](https://github.com/cypress-io/cypress/pull/25992) so I wonder if once that is done the badge will stop flickering, or if it will just flicker to no badge. Anyway, just something I noticed in initial testing

**Comment by lmiller1990** [comment]
I am still going over the code, but this worked well when I tested it. I did real world testing, against an actual production app where I use Cypress for CT and E2E. 

I did have to change from and back to the tab to see it update real time - I was just staring at it, expecting it to change - I guess the subscription only triggers if there is a run in progress when you open the page?

Watching it real time was really good - I could start debugging my tests (failed ones) without waiting! Very useful.

**Comment by warrensplayer** [comment]
> I did have to change from and back to the tab to see it update real time - I was just staring at it, expecting it to change - I guess the subscription only triggers if there is a run in progress when you open the page?

@lmiller1990  I saw this when developing but then was not able to reproduce it.  I will try this scenario again and see if I can make it happen.

**Comment by warrensplayer** [comment]
> One thing I'm noticing in initial testing is that there's some flickering back and forth with the navigation badge (seemingly when new data comes in from the cloud?) - I'll see the number of failures, then the badge will flash to "New" for a second, and then back to the number of failures (this time updated since more tests are failing).

@astone123 I have also seen this happen.  It seems to only happen once for me and then stop acting that way. If you are able to consistently see it, please try to recall the steps you took to get it to happen.

**Comment by warrensplayer** [comment]
Feedback from acceptance testing with Product and Design
- ~add “previous” to banner~ **DONE**
- ~banner hide delay fix~ **DONE**
- ~run persistence fix~ **DONE**

**Comment by warrensplayer** [comment]
> > I did have to change from and back to the tab to see it update real time - I was just staring at it, expecting it to change - I guess the subscription only triggers if there is a run in progress when you open the page?
> 
> @lmiller1990 I saw this when developing but then was not able to reproduce it. I will try this scenario again and see if I can make it happen.

@lmiller1990 I tested again and I think this is just a matter of waiting for the next outer poll to happen which is every 30 seconds.  The "outer" poll is the one from `RelevantRunsDataSource` that is sending the Git commit shas and determining the appropriate relevant runs.  If a run starts recording a split second after one poll, you would have to wait an entire 30 seconds before it polls again.  Try your test again and see if that is the case.  There are DEBUG logs in the datasources you can turn on to see when it polls.

File: packages/app/src/debug/DebugContainer.cy.tsx [hunk 1]
```diff
@@ -281,37 +283,151 @@ describe('<DebugContainer />', () => {
...
+        newTest.id = '345'
+        newTest.thumbprint = `${newTest.thumbprint}c`
+
+        run.testsForReview.push(newTest)
+      })
+
+      cy.wait(waitTimeBetweenSimulatedEvents)
+      cy.findAllByTestId('debug-spec-item').should('have.length', 1)
+      cy.get<CloudRun>('@run').then((run) => {
+        const newSpec: CloudSpecRun = JSON.parse(JSON.stringify(run.specs[0]))
```
Comment (lmiller1990) [inline packages/app/src/debug/DebugContainer.cy.tsx:377]: This is fine - no change request - but just out of interested, I found out about `window.structuredClone` recently, which is similar to this parse/stringify trick: https://developer.mozilla.org/en-US/docs/Web/API/structuredClone

Neat little function!

File: packages/data-context/src/sources/RelevantRunsDataSource.ts [hunk 2]
```diff
@@ -181,16 +181,22 @@ export class RelevantRunsDataSource {
...
+  async moveToNext (runNumber: number, shas: string[]) {
     debug('Moving to next relevant run')
 
-    await this.checkRelevantRuns(shas, true)
+    await this.checkRelevantRuns(shas, true, runNumber)
   }
 
-  async checkRelevantRuns (shas: string[], clearCache: boolean = false) {
+  /**
+   * TODO Document this function
```
Comment (lmiller1990) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts]: Should document this at some point 🤔

File: packages/data-context/test/unit/sources/RelevantRunSpecsDataSource.spec.ts [hunk 3]
```diff
@@ -51,18 +60,85 @@ describe('RelevantRunSpecsDataSource', () => {
...
         statuses: {
           current: 'PASSED',
           next: 'RUNNING',
         },
+        testCounts: { current: 7, next: 0 },
+      })
+    })
+  })
+
+  describe('polling', () => {
```
Comment (lmiller1990) [inline packages/data-context/test/unit/sources/RelevantRunSpecsDataSource.spec.ts:81]: Neat way to test this 👍

**Review by lmiller1990 — APPROVED** [review]
This is looking good to me - I'll do a little more testing but the code seems mostly straight forward.

File: packages/data-context/src/sources/RelevantRunsDataSource.ts (hunk 2 above)
Comment (warrensplayer) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts]: Done

File: packages/app/src/debug/DebugContainer.cy.tsx (hunk 1 above)
Comment (warrensplayer) [reply packages/app/src/debug/DebugContainer.cy.tsx:377]: I had not seen that before.  I will check it out.  I just needed something basic to make sure I could mutate the test data without affecting other tests.

File: cli/CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,7 @@ _Released 03/14/2023 (PENDING)_
 **Features:**
 
 - It is now possible to control the number of connection attempts to the browser using the CYPRESS_CONNECT_RETRY_THRESHOLD Environment Variable. Learn more [here](https://docs.cypress.io/guides/references/advanced-installation#Environment-variables). Addressed in [#25848](https://github.com/cypress-io/cypress/pull/25848).
+- The Debug page is now able to show real-time results from runs that are still running in Cypress Cloud.  Addresses [#25759](https://github.com/cypress-io/cypress/issues/25759).
```
Comment (mike-plummer) [inline cli/CHANGELOG.md]:
```suggestion
- The Debug page is now able to show real-time results from in-progress runs.  Addresses [#25759](https://github.com/cypress-io/cypress/issues/25759).
```

Using "run" twice feels a bit awkward, just a wording suggestion

File: packages/app/src/debug/DebugContainer.vue [hunk 5]
```diff
@@ -22,19 +22,25 @@
...
           :gql="run"
           :commits-ahead="props.commitsAhead"
         />
+        <TransitionQuickFade>
+          <DebugTestingProgress v-if="isRunning" />
+        </TransitionQuickFade>
+
         <DebugPendingRunSplash
-          v-if="isFirstPendingRun"
+          v-if="isRunning && (!run.totalFailed || run.totalFailed === 0)"
```
Comment (mike-plummer) [inline packages/app/src/debug/DebugContainer.vue]: I think this is overkill - wouldn't `!run.totalFailed` cover the `0` case as well?

File: packages/app/src/debug/DebugContainer.vue [hunk 6]
```diff
@@ -221,9 +229,11 @@ const debugSpecsArray = computed(() => {
   return []
 })
 
-const newerRelevantRun = computed(() => nextRun.value)
+const nextRelevantRun = computed(() => nextRun.value)
+
+const isRunning = computed(() => !!run.value && run.value.status === 'RUNNING')
 
-const isFirstPendingRun = computed(() => run.value && run.value.status === 'RUNNING')
+const isScheduledToComplete = computed(() => !!run.value && !!run.value.scheduledToCompleteAt)
```
Comment (mike-plummer) [inline packages/app/src/debug/DebugContainer.vue]:
```suggestion
const isScheduledToComplete = computed(() => !!run.value?.scheduledToCompleteAt)
```

File: packages/app/src/debug/DebugNewRelevantRunBar.vue [hunk 7]
```diff
@@ -92,15 +97,21 @@ const specCounts = computed(() => {
   return specs.data.value?.relevantRunSpecChange?.currentProject?.relevantRunSpecs?.next
 })
 
+const isPrevious = computed(() => {
+  return props.currentRunNumber && props.gql.runNumber && props.gql.runNumber < props.currentRunNumber
```
Comment (mike-plummer) [inline packages/app/src/debug/DebugNewRelevantRunBar.vue]: Just to sanity check - are run numbers 0-indexed or 1-indexed?

File: packages/app/src/debug/DebugPageHeader.vue [hunk 8]
```diff
@@ -156,7 +158,31 @@ const relativeCreatedAt = computed(() => dayjs(new Date(debug.value.createdAt!))
...
   Ex: 1 hour and 1 second which is 3601000ms = 01h 00m 01s
 */
-const totalDuration = computed(() => dayjs.duration(debug.value.totalDuration ?? 0).format('HH[h] mm[m] ss[s]').replace(/^0+h /, ''))
+const formatDuration = (duration: number) => {
+  return dayjs.duration(duration).format('HH[h] mm[m] ss[s]').replace(/^0+h /, '')
+}
+
+const formatCreatedAt = (createdAt: string) => {
+  return dayjs(new Date(createdAt)).fromNow()
+}
```
Comment (mike-plummer) [inline packages/app/src/debug/DebugPageHeader.vue:167]: Could extract these as composable logic alongside `useDurationFormat` to make them easier to test

File: packages/graphql/test/stubCloudTypes.ts [hunk 9]
```diff
@@ -533,3 +535,50 @@ export const CloudQuery: MaybeResolver<Query> = {
...
+    scheduledToCompleteAt,
+  }
+
+  if (type === 'current') {
+    event.relevantRunSpecChange.currentProject.relevantRunSpecs.current = relevantRunSpec
+  } else {
+    event.relevantRunSpecChange.currentProject.relevantRunSpecs.next = relevantRunSpec
+  }
+
+  console.log('returning event', event)
```
Comment (mike-plummer) [inline packages/graphql/test/stubCloudTypes.ts]: Remove?

File: packages/graphql/test/stubCloudTypes.ts [hunk 10]
```diff
@@ -533,3 +535,50 @@ export const CloudQuery: MaybeResolver<Query> = {
...
+
+// eslint-disable-next-line no-redeclare
+export function createRelevantRunSpecChangeEvent (type: 'next', completed: number, total: number, scheduledToCompleteAt: string | null): DebugNewRelevantRunBar_SpecsSubscription
+
+// eslint-disable-next-line no-redeclare
+export function createRelevantRunSpecChangeEvent (type: 'next', completed: number, total: number): DebugNewRelevantRunBar_SpecsSubscription
+
+// eslint-disable-next-line no-redeclare
+export function createRelevantRunSpecChangeEvent (type: 'current' | 'next', completed: number, total: number, scheduledToCompleteAt: string | null = null): EventType {
+  const event: any = {
```
Comment (mike-plummer) [inline packages/graphql/test/stubCloudTypes.ts:554]: `any` necessary here? Function has a return type this evidently conforms to

File: packages/graphql/src/schemaTypes/objectTypes/gql-RelevantRun.ts [hunk 11]
```diff
@@ -33,5 +33,9 @@ export const RelevantRunSpecs = objectType({
     t.nonNull.int('completedSpecs', {
       description: 'Number of specs in the run that have finished being processed',
     })
+
+    t.dateTime('scheduledToCompleteAt', {
```
Comment (mike-plummer) [inline packages/graphql/src/schemaTypes/objectTypes/gql-RelevantRun.ts:37]: Just curious re: use of `t.dateTime` - other object schemas seem to use `string` with an ISO-8601 timestamp value. Does `dateTime` auto-parse into a `Date`?

File: packages/data-context/src/sources/RelevantRunsDataSource.ts [hunk 12]
```diff
@@ -181,16 +181,24 @@ export class RelevantRunsDataSource {
   /**
    * Clear the cached current run to allow the data source to pick the next completed run as the current
    */
-  async moveToNext (shas: string[]) {
+  async moveToNext (runNumber: number, shas: string[]) {
```
Comment (mike-plummer) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts]: Should this still be named `moveToNext`? Seems like it's more of a `moveToRun` now

File: packages/graphql/test/stubCloudTypes.ts (hunk 10 above)
Comment (warrensplayer) [reply packages/graphql/test/stubCloudTypes.ts:554]: There may be a better way to type this, but without `any` then other type errors happen below this

File: packages/graphql/src/schemaTypes/objectTypes/gql-RelevantRun.ts (hunk 11 above)
Comment (warrensplayer) [reply packages/graphql/src/schemaTypes/objectTypes/gql-RelevantRun.ts:37]: I was matching the type coming from the Cloud schema.  It still ends up as a string in the generated types.

File: packages/data-context/src/sources/RelevantRunsDataSource.ts (hunk 12 above)
Comment (warrensplayer) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts]: Good call!

File: packages/graphql/test/stubCloudTypes.ts (hunk 9 above)
Comment (warrensplayer) [reply packages/graphql/test/stubCloudTypes.ts]: Done

File: packages/app/src/debug/DebugNewRelevantRunBar.vue (hunk 7 above)
Comment (warrensplayer) [reply packages/app/src/debug/DebugNewRelevantRunBar.vue]: Added a better check

File: packages/app/src/debug/DebugContainer.vue (hunk 5 above)
Comment (warrensplayer) [reply packages/app/src/debug/DebugContainer.vue]: Yes it would.  Fixed

File: packages/app/src/debug/DebugPageHeader.vue (hunk 8 above)
Comment (warrensplayer) [reply packages/app/src/debug/DebugPageHeader.vue:167]: I decided to just leave these here for now since they are only used here.  I added some tests in the CT tests to cover the scenarios.

**Review by mike-plummer — APPROVED** [review]
LGTM. Only remaining feedback is that I would like to see a few Percy snapshots added for the new UI behaviors, but that's non-blocking

File: packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts [hunk 13]
```diff
@@ -147,5 +152,41 @@ describe('RelevantRunsDataSource', () => {
         commitsAhead: 0,
       })
     })
+
+    it('preserves running if switched', async () => {
```
Comment (lmiller1990) [inline packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts:156]: What does "switching" imply here - when you click on "Change to this run" (or whatever the button says)?

I'm guessing the reason no-one found this edge case in testing was because it was closely related to the timing of when you change runs?

Hope we can find a good way to E2E test this at some point - unit tests do the job, but I imagine it'd be hard to fully grok what this does from a user point of view without an E2E test.

**Review by lmiller1990 — APPROVED** [review]

**Comment by lmiller1990** [comment]
Let's merge it! Asked a question around the latest commit but it's just a question, not a blocker on merging.
